### PR TITLE
Fix issue #1258

### PR DESCRIPTION
**Pull request by elcore**
Hello @mholt ,

this PR should fix issue #1258:

```
2016/11/24 16:10:41 [INFO] Obtaining new certificate for 1.caddy.domain.com
2016/11/24 16:10:42 [INFO] acme: Registering account for
2016/11/24 16:10:42 [INFO][1.caddy.domain.com] acme: Obtaining bundled SAN certificate
2016/11/24 16:10:43 [INFO][1.caddy.domain.com] acme: Trying to solve TLS-SNI-01
2016/11/24 16:10:45 [INFO][1.caddy.domain.com] The server validated our request
2016/11/24 16:10:45 [INFO][1.caddy.domain.com] acme: Validations succeeded; requesting certificates
2016/11/24 16:10:46 [INFO] acme: Requesting issuer cert from https://acme-staging.api.letsencrypt.org/acme/issuer-cert
2016/11/24 16:10:46 [INFO][1.caddy.domain.com] Server responded with a certificate.
2016/11/24 16:10:46 [WARNING] Stapling OCSP: no OCSP stapling for [1.caddy.domain.com]: ocsp: error from server: unauthorized
2016/11/24 16:10:53 [INFO] Obtaining new certificate for 2.caddy.domain.com
2016/11/24 16:10:53 [INFO][2.caddy.domain.com] acme: Obtaining bundled SAN certificate
2016/11/24 16:10:53 [INFO][2.caddy.domain.com] acme: Could not find solver for: dns-01
2016/11/24 16:10:53 [INFO][2.caddy.domain.com] acme: Trying to solve HTTP-01
2016/11/24 16:10:54 [INFO][2.caddy.domain.com] Served key authentication
2016/11/24 16:10:55 [INFO][2.caddy.domain.com] The server validated our request
2016/11/24 16:10:55 [INFO][2.caddy.domain.com] acme: Validations succeeded; requesting certificates
2016/11/24 16:10:56 [INFO] acme: Requesting issuer cert from https://acme-staging.api.letsencrypt.org/acme/issuer-cert
2016/11/24 16:10:56 [INFO][2.caddy.domain.com] Server responded with a certificate.
2016/11/24 16:10:56 [WARNING] Stapling OCSP: no OCSP stapling for [2.caddy.domain.com]: ocsp: error from server: unauthorized
2016/11/24 16:11:04 [INFO] Obtaining new certificate for 3.caddy.domain.com
2016/11/24 16:11:04 [INFO][3.caddy.domain.com] acme: Obtaining bundled SAN certificate
2016/11/24 16:11:04 [INFO][3.caddy.domain.com] acme: Trying to solve TLS-SNI-01
2016/11/24 16:11:06 [INFO][3.caddy.domain.com] The server validated our request
2016/11/24 16:11:06 [INFO][3.caddy.domain.com] acme: Validations succeeded; requesting certificates
2016/11/24 16:11:07 [INFO] acme: Requesting issuer cert from https://acme-staging.api.letsencrypt.org/acme/issuer-cert
2016/11/24 16:11:07 [INFO][3.caddy.domain.com] Server responded with a certificate.
2016/11/24 16:11:07 [WARNING] Stapling OCSP: no OCSP stapling for [3.caddy.domain.com]: ocsp: error from server: unauthorized
2016/11/24 16:11:40 [INFO] Obtaining new certificate for 4.caddy.domain.com
2016/11/24 16:11:40 [INFO][4.caddy.domain.com] acme: Obtaining bundled SAN certificate
2016/11/24 16:11:40 [INFO][4.caddy.domain.com] acme: Trying to solve HTTP-01
2016/11/24 16:11:41 [INFO][4.caddy.domain.com] Served key authentication
2016/11/24 16:11:42 [INFO][4.caddy.domain.com] The server validated our request
2016/11/24 16:11:42 [INFO][4.caddy.domain.com] acme: Validations succeeded; requesting certificates
2016/11/24 16:11:42 [INFO] acme: Requesting issuer cert from https://acme-staging.api.letsencrypt.org/acme/issuer-cert
2016/11/24 16:11:42 [INFO][4.caddy.domain.com] Server responded with a certificate.
```

### CODE REVIEW

**Comment by elcore** [comment]
Now better @mholt ?

**Comment by mholt** [comment]
Now that's more interesting. Why does this work? (And can you write a test for it?) What was causing the problem?

**Comment by elcore** [comment]
> Now that's more interesting. Why does this work? (And can you write a test for it?) What was causing the problem?

https://github.com/mholt/caddy/blob/master/caddy.go#L229-#L231
```
[...] as listeners tend to report
// [::]:80, for example, when the matching address that
// created the listener might be simply :80.
```

```
--- FAIL: TestListenerAddrEqual (0.00s)
        caddy_test.go:82: Test 0 (127.0.0.1:2016 == [::]:2016): expected true but was false
FAIL
FAIL    github.com/mholt/caddy  0.026s
```

**Comment by elcore** [comment]
Everything okay @mholt? -- Is there something I need to change?

**Comment by mholt** [comment]
I'll take a look after the holiday

**Comment by elcore** [comment]
> I'll take a look after the holiday

Sure, enjoy it :)

File: caddy.go [hunk 1]
```diff
@@ -241,6 +241,9 @@ func listenerAddrEqual(ln net.Listener, addr string) bool {
 	if lnAddr == net.JoinHostPort("0.0.0.0", port) {
 		return true
 	}
+	if hostname != "" && lnAddr == addr {
+		return true
+	}
 	return false
```
Comment (mholt) [inline caddy.go]: These last 4 lines can be replaced with simply: `return hostname != "" && lnAddr == addr`

File: caddy_test.go [hunk 2]
```diff
@@ -56,3 +59,38 @@ func TestIsLoopback(t *testing.T) {
 		}
 	}
 }
+
+func TestListenerAddrEqual(t *testing.T) {
+	ln1, err := newLocalListener("[::]:2016")
```
Comment (mholt) [inline caddy_test.go]: I use ports 2017 and 2018 in development sometimes (silly, I know) -- would probably be better to use port 0 and then use ln.Address() to get the address of the listener. Using port 0 will let the system assign a port that's available.

File: caddy_test.go [hunk 3]
```diff
@@ -56,3 +59,38 @@ func TestIsLoopback(t *testing.T) {
...
+	}
+	defer ln2.Close()
+
+	for i, test := range []struct {
+		ln     net.Listener
+		addr   string
+		expect bool
+	}{
+		{ln1, "0.0.0.0:2016", true},
+		{ln2, "0.0.0.0:2018", false},
```
Comment (mholt) [inline caddy_test.go]: Can we add some more test cases here? Use the test coverage tool to ensure all lines of the `listenerAddrEqual()` function are being hit.

File: caddy_test.go [hunk 4]
```diff
@@ -56,3 +59,38 @@ func TestIsLoopback(t *testing.T) {
...
+		}
+	}
+}
+
+func newLocalListener(addr string) (net.Listener, error) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return ln, nil
```
Comment (mholt) [inline caddy_test.go]: This whole function can become `return net.Listen("tcp", addr)` which, in turn, doesn't need to be a separate function at all. :)

File: caddy_test.go [hunk 5]
```diff
@@ -56,3 +59,38 @@ func TestIsLoopback(t *testing.T) {
...
+	for i, test := range []struct {
+		ln     net.Listener
+		addr   string
+		expect bool
+	}{
+		{ln1, "0.0.0.0:2016", true},
+		{ln2, "0.0.0.0:2018", false},
+	} {
+		if got, want := listenerAddrEqual(test.ln, test.addr), test.expect; got != want {
+			t.Errorf("Test %d (%v == %v): expected %v but was %v", i, test.addr, test.ln.Addr().String(), want, got)
```
Comment (mholt) [inline caddy_test.go]: Use %s for the string types

**Comment by mholt** [comment]
(I should figure out a way to add test coverage to CI) Do these tests cover every branch of the function?

**Comment by elcore** [comment]
Hello @mholt ,

no! It does not cover

```
if err != nil { //covered
                return lnAddr == addr // not covered
} //covered

if lnAddr == net.JoinHostPort("0.0.0.0", port) {  //covered
                return true // not covered
} //covered
```

I will create a new PR, you where a little bit to fast 😄

**Pull request merged**